pkg/pipelines/webhook/webhook.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/openshift/odo/pkg/pipelines/routes"
+	"github.com/openshift/odo/pkg/pipelines/secrets"
+)
+
+type webhookInfo struct {
+	clusterResource *resources
+	repository      *git.Repository
+	gitRepoURL      string
+	cicdNamepace    string
+	listenerURL     string
+	accessToken     string
+	names           []string
+	isCICD          bool
+}
+
+// Create creates a new webhook on the target Git Repository
+// names is a {envName, appName, seviceName} tuple.
+// It returns the ID of created webhook.
+func Create(accessToken, pipelinesFile string, names []string, isCICD bool) (string, error) {
+
+	webhook, err := newWebhookInfo(accessToken, pipelinesFile, names, isCICD)
+	if err != nil {
+		return "", err
+	}
+
+	exists, err := webhook.exists()
+	if err != nil {
+		return "", err
+	}
+
+	if exists {
+		return "", errors.New("webhook already exists")
+	}
+
+	return webhook.create()
+}
+
+// Delete deletes webhooks on the target Git Repository that match the listener address
+// names is a {envName, appName, seviceName} tuple.
+// It returns the IDs of deleted webhooks.
+func Delete(accessToken, pipelinesFile string, names []string, isCICD bool) ([]string, error) {
+
+	webhook, err := newWebhookInfo(accessToken, pipelinesFile, names, isCICD)
+	if err != nil {
+		return nil, err
+	}
+
+	ids, err := webhook.list()
+	if err != nil {
+		return nil, err
+	}
+
+	return webhook.delete(ids)
+}
+
+// List returns an array of webhook IDs for the target Git repository/listeners
+func List(accessToken, pipelinesFile string, names []string, isCICD bool) ([]string, error) {
+
+	webhook, err := newWebhookInfo(accessToken, pipelinesFile, names, isCICD)
+	if err != nil {
+		return nil, err
+	}
+
+	return webhook.list()
+}
+
+func newWebhookInfo(accessToken, pipelinesFile string, names []string, isCICD bool) (*webhookInfo, error) {
+
+	manifest, err := config.ParseFile(ioutils.NewFilesystem(), pipelinesFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse pipelines: %w", err)
+	}
+
+	if err := manifest.Validate(); err != nil {
+		return nil, err
+	}
+
+	gitRepoURL := getRepoURL(manifest, isCICD, names)
+	if gitRepoURL == "" {
+		return nil, errors.New("failed to find Git repostory URL in manifest")
+	}
+
+	cicdNamepace := getCICDNamespace(manifest)
+	if cicdNamepace == "" {
+		return nil, errors.New("failed to find CICD namespace in manifest")
+	}
+
+	clusterResources, err := newResources()
+	if err != nil {
+		return nil, err
+	}
+
+	repository, err := git.NewRepository(gitRepoURL, accessToken)
+	if err != nil {
+		return nil, err
+	}
+
+	listenerURL, err := getListenerURL(clusterResources, cicdNamepace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get event listener URL: %w", err)
+	}
+
+	return &webhookInfo{clusterResources, repository, gitRepoURL, cicdNamepace, listenerURL, accessToken, names, isCICD}, nil
+}
+
+func (w *webhookInfo) exists() (bool, error) {
+
+	ids, err := w.repository.ListWebhooks(w.listenerURL)
```

Comment:
Couldn't this delegate via `list()` ?

Reply:
There is nothing wrong with that.   It is cleaner apis for the caller which only cares if object exists and careless about the list items.  That's why containers implement isEmpty/exists methods.

Reply:
I'm not sure you understood my point, both `exists` and `list` call `w.repository.ListWebhooks(w.listenerURL)`, `exists` could call `list` and return the `len()` of the result.

Reply:
OK.  I see what you are saying.   Your proposal is slighty less efficient as there is an extra function call.   It does not make the code cleaner.   I am not sure I see any benefit of doing so.

Reply:
It would avoid duplicating the code.

Reply:
Calling the a same method in two different functions for different reasons is not a duplicate piece of code.    Copying a block of code and repeat that block of code in multiple places instead of making it a function and call it is duplicating of code.    You just make a method call twice unnecessarily.